tests/virt/cluster/common_templates/utils.py

```diff
@@ -628,3 +643,76 @@ def assert_windows_efi(vm):
 @cache
 def is_jira_64776_bug_open():
     return is_jira_open(jira_id="CNV-64776")
+
+
+def get_matrix_os_golden_image_data_source(
+    admin_client: DynamicClient, golden_images_namespace: Namespace, os_matrix: dict[str, dict]
+) -> Generator[DataSource, None, None]:
+    """Retrieves or creates a DataSource object in golden image namespace specified in the OS matrix.
+
+    Args:
+        admin_client (DynamicClient): Kubernetes dynamic client.
+        golden_images_namespace (Namespace): Namespace where golden images are stored.
+        os_matrix (dict[str, dict]): *_os_matrix dict
+
+    Yields:
+        DataSource: DataSource object.
+    """
+
+    os_name = [*os_matrix][0]
+    os_dict = os_matrix[os_name]
+    data_source_name = os_dict[DATA_SOURCE_STR]
+
+    data_source = DataSource(client=admin_client, name=data_source_name, namespace=golden_images_namespace.name)
+    if data_source.exists and data_source.instance.status.source:
+        LOGGER.info(f"DataSource {data_source_name} already exists and has a source pvc/snapshot.")
+        yield data_source
+    else:
+        LOGGER.warning(f"No DataSource {data_source_name} found or it doesn't have a source pvc/snapshot.")
+        with create_dv(
+            dv_name=os_name,
+            namespace=golden_images_namespace.name,
+            storage_class=py_config["default_storage_class"],
+            url=f"{get_test_artifact_server_url()}{os_dict['image_path']}",
+            size=os_dict["dv_size"],
+            client=admin_client,
+        ) as dv:
+            dv.wait_for_dv_success(timeout=TIMEOUT_30MIN)
+            yield from create_or_update_data_source(admin_client=admin_client, dv=dv)
```

Comment:
With this fallback, if DataSources are broken, we may not notice, and we will still think we are creating VMs from them

Reply:
broken how?
currently there are datasources with non-existing source (in which case we create new dv and update the DS CR to use new source)

Reply:
Like if `DataSource` object exists, but doesn't have a `status.source` even if it supposed to have it.

And it will be hard to track which tests are using boot source images, and which are creating the source themselves...

Reply:
> Like if `DataSource` object exists, but doesn't have a `status.source` even if it supposed to have it.
> 
i don't think this is the place to test DataSoruces and what they should have as source.

> And it will be hard to track which tests are using boot source images, and which are creating the source themselves...
>
you'll see it by exec time (obviously creating dv takes longer then clone from existing) and from console logs.
also i don't think it really matters for tester where exactly VM will get disk from (except storage team obviously)
if we want to make sure all pre-existing datasources are with valid source, i believe it should be a dedicated test (maybe in cluster healthcheck?).

Reply:
then for the os matrix support we're ok.
i believe there are dataSource-related tests to make sure that the required PVCs exist.